### ISSUE TITLE
Test_autocmd_bufreadpre when terminal has >= 164 lines

### ISSUE THREAD

**Issue by dpelle** [vim/vim#5260]
`Test_autocmd_bufreadpre` fails when terminal has >= 164 lines.

**To Reproduce**

1. Create a terminal with many lines (>= 164 lines, command `:echo &lines` should report 164 or more). You may need to reduce font size for this and use a big screen.
2. run `make test_autocmd` and observe that it fails:
```
$ make test_autocmd
rm -f test_autocmd.res test.log messages
VIMRUNTIME=../../runtime  ../vim -f  -u unix.vim -U NONE --noplugin --not-a-term -S runtest.vim test_autocmd.vim --cmd 'au SwapExists * let v:swapchoice = "e"' > /dev/null


From test_autocmd.vim:
Executed Test_Acmd_BufAll() in   0.015032 seconds
Executed Test_Acmd_BufEnter() in   0.006295 seconds
Executed Test_BufEnter() in   0.000519 seconds
Executed Test_BufLeave_Wipe() in   0.008668 seconds
Executed Test_BufReadCmdHelp() in   0.007064 seconds
Executed Test_BufReadCmdHelpJump() in   0.015155 seconds
Executed Test_BufWritePre() in   0.007357 seconds
Executed Test_BufWrite_lockmarks() in   0.018033 seconds
Executed Test_BufleaveWithDelete() in   0.000693 seconds
Executed Test_ChangedP() in   0.002236 seconds
Executed Test_Changed_FirstTime() in   0.147347 seconds
Executed Test_Cmd_Autocmds() in   0.008320 seconds
Executed Test_Cmdline() in   0.002745 seconds
Executed Test_Filter_noshelltemp() in   0.016074 seconds
Executed Test_OptionSet() in   0.010360 seconds
Executed Test_OptionSet_diffmode() in   0.001783 seconds
Executed Test_OptionSet_diffmode_close() in   0.001536 seconds
Executed Test_OptionSet_modeline() in   0.004073 seconds
Executed Test_QuitPre() in   0.001588 seconds
Executed Test_ReadWrite_Autocmds() in   0.071662 seconds
Executed Test_TextChangedI_with_setline() in   0.001664 seconds
Executed Test_TextYankPost() in   0.002514 seconds
Executed Test_ai_CTRL_O() in   0.001109 seconds
Executed Test_augroup_deleted() in   0.000554 seconds
Executed Test_augroup_warning() in   0.000831 seconds
Executed Test_autocmd_CmdWinEnter() in   0.098720 seconds
Executed Test_autocmd_SafeState() in   0.111318 seconds
Executed Test_autocmd_bufreadpre() in   0.007719 seconds
Executed Test_autocmd_bufunload_avoiding_SEGV_01() in   0.001647 seconds
Executed Test_autocmd_bufunload_avoiding_SEGV_02() in   0.001207 seconds
Executed Test_autocmd_bufunload_with_tabnext() in   0.000723 seconds
Executed Test_autocmd_bufwinleave_with_tabfirst() in   0.001394 seconds
Executed Test_autocmd_bufwipe_in_SessLoadPost() in   1.075588 seconds
Executed Test_autocmd_bufwipe_in_SessLoadPost2() in   0.041916 seconds
Executed Test_autocmd_nested() in   0.000913 seconds
Executed Test_autocmd_once() in   0.000249 seconds
Executed Test_autocmd_was_using_freed_memory() in   0.000544 seconds
Executed Test_buflocal_autocmd() in   0.000858 seconds
Executed Test_bufunload() in   0.000424 seconds
Executed Test_bufunload_all() in   0.037034 seconds
Executed Test_change_mark_in_autocmds() in   0.035193 seconds
Executed Test_cursorhold_insert() in   0.100236 seconds
Executed Test_cursorhold_insert_ctrl_x() in   0.099993 seconds
Executed Test_cursorhold_insert_with_timer_interrupt() in   1.011690 seconds
Executed Test_dirchanged_auto() in   0.001779 seconds
Executed Test_dirchanged_global() in   0.000831 seconds
Executed Test_dirchanged_local() in   0.000846 seconds
Executed Test_early_bar() in   0.000675 seconds
Executed Test_empty_doau() in   0.000141 seconds
Executed Test_nocatch_wipe_all_buffers() in   0.000665 seconds
Executed Test_nocatch_wipe_dummy_buffer() in   0.000450 seconds
Executed Test_three_windows() in   0.023116 seconds
Executed Test_throw_in_BufWritePre() in   0.000849 seconds
Executed Test_vim_did_enter() in   0.000236 seconds
Executed Test_win_tab_autocmd() in   0.001002 seconds
Executed 55 tests in   3.024975 seconds
1 FAILED:
Found errors in Test_autocmd_bufreadpre():
function RunTheTest[40]..Test_autocmd_bufreadpre line 16: Expected range 49 - 51, but got 48
Makefile:75: recipe for target 'test_autocmd' failed
make: *** [test_autocmd] Error 1
```

The failing test assert is at test_autocmd.vim:1979:
```
  1976   " triggers BufReadPre, should not move the cursor in either window
  1977   " The topline may change one line in a large window.
  1978   edit
!!1979   call assert_inrange(g:wsv2.topline - 1, g:wsv2.topline + 1, winsaveview().topline)
```
I don't understand why the topline changes here. Notice the comment
at line 1977 that says that topline may change by one line for a large
window. I suppose that for a huge window, it may change by even more.

**Environment**
 - vim-8.1.2337 (huge)
 - xubuntu-18.04.3
 - xfce4-terminal 0.8.7.4

**Comment by brammool** [comment]
Dominique wrote:

> `Test_autocmd_bufreadpre` fails when terminal has >= 164 lines.
> 
> **To Reproduce**
> 
> 1. Create a terminal with many lines (>= 164 lines, command `:echo &lines` should report 164 or more). You may need to reduce font size for this and use a big screen.
> 2. run `make test_autocmd` and observe that it fails:
> ```
> $ make test_autocmd
> rm -f test_autocmd.res test.log messages
> VIMRUNTIME=../../runtime  ../vim -f  -u unix.vim -U NONE --noplugin --not-a-term -S runtest.vim test_autocmd.vim --cmd 'au SwapExists * let v:swapchoice = "e"' > /dev/null
> 
> 
> >From test_autocmd.vim:
[...]
> 1 FAILED:
> Found errors in Test_autocmd_bufreadpre():
> function RunTheTest[40]..Test_autocmd_bufreadpre line 16: Expected range 49 - 51, but got 48
> Makefile:75: recipe for target 'test_autocmd' failed
> make: *** [test_autocmd] Error 1
> ```
> 
> The failing test assert is at test_autocmd.vim:1979:
> ```
>   1976   " triggers BufReadPre, should not move the cursor in either window
>   1977   " The topline may change one line in a large window.
>   1978   edit
> !!1979   call assert_inrange(g:wsv2.topline - 1, g:wsv2.topline + 1, winsaveview().topline)
> ```
> I don't understand why the topline changes here. Notice the comment
> at line 1977 that says that topline may change by one line for a large
> window. I suppose that for a huge window, it may change by even more.
> 
> **Environment**
>  - vim-8.1.2337 (huge)
>  - xubuntu-18.04.3
>  - xfce4-terminal 0.8.7.4

I can't reproduce it right now.  Perhaps it helps if the setline() fills
more lines with text?

-- 
hundred-and-one symptoms of being an internet addict:
114. You are counting items, you go "0,1,2,3,4,5,6,7,8,9,A,B,C,D...".

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
Fixed in patch [8.2.0432](https://github.com/vim/vim/releases/tag/v8.2.0432).
